File: LeetCode/212.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        def helper(r, c, node, chars=""):
            if (r, c) in self.seen:
                return

            if board[r][c] in node.children:
                char = board[r][c]
                nxt = node.children[char]
                self.seen.add((r, c))

                if r > 0:
                    helper(r - 1, c, nxt, chars+char)
                if r < self.rows - 1:
                    helper(r + 1, c, nxt, chars+char)
                if c > 0:
                    helper(r, c - 1, nxt, chars+char)
                if c < self.cols - 1:
                    helper(r, c + 1, nxt, chars+char)

                if nxt.isWord:
                    self.res.add(chars+char)
                self.seen.remove((r, c))

            return

        self.root = TrieNode()
        self.res = set()
        self.rows = len(board)
        self.cols = len(board[0])
        self.seen = set()

        for word in words:
            self.curr = self.root
            for char in word:
                if char not in self.curr.children:
                    self.curr.children[char] = TrieNode()
                self.curr = self.curr.children[char]
            self.curr.isWord = True
        
        for r in range(self.rows):
            for c in range(self.cols):
                helper(r, c, self.root)

        return list(self.res)

File: leetcode-new/212.cpp
class TrieNode
{
public:
    unordered_map<char, TrieNode *> children;
    bool isEnd;
    TrieNode() : children{}, isEnd{false} {}

    void add(string word)
    {
        TrieNode *curr = this;
        for (char c : word)
        {
            auto &wl = curr->children;
            if (wl.find(c) == wl.end())
                wl[c] = new TrieNode();
            curr = wl[c];
        }
        curr->isEnd = true;
    }
};

class Solution
{
public:
    vector<string> findWords(vector<vector<char>> &board, vector<string> &words)
    {
        TrieNode *root = new TrieNode();
        this->board = board;
        r = board.size();
        c = board[0].size();
        for (auto word : words)
            root->add(word);

        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < c; j++)
            {
                helper(i, j, root, "");
            }
        }
        return res;
    }

private:
    vector<vector<char>> board;
    vector<string> res{};
    int r, c;
    vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    void helper(int cr, int cc, TrieNode *node, string word)
    {
        if (cr < 0 || cr >= r || cc < 0 || cc >= c || board[cr][cc] == '#' || node->children.find(board[cr][cc]) == node->children.end())
            return;

        char c = board[cr][cc];
        board[cr][cc] = '#';
        node = node->children[c];
        word += c;
        if (node->isEnd)
        {
            node->isEnd = false;
            res.push_back(word);
        }

        for (const auto &[dx, dy] : dirs)
        {
            helper(cr + dx, cc + dy, node, word);
        }
        board[cr][cc] = c;
    }
};
